iwyu.cc

```diff
@@ -1700,8 +1701,25 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       ReportTypeUse(used_loc, type);
   }
 
+  void ReportNamespaceDeclUse(SourceLocation used_loc,
+                              const NamespaceDecl* namespace_decl,
+                              const char* comment = nullptr,
+                              UseFlags extra_use_flags = 0) {
+    // Canonicalize the use location and report the use.
+    used_loc = GetCanonicalUseLocation(used_loc, namespace_decl);
+    const FileEntry* used_in = GetFileEntry(used_loc);
+    preprocessor_info().FileInfoFor(used_in)->ReportNamespaceDeclUse(
+        used_loc, namespace_decl, extra_use_flags, comment);
+  }
+
   //------------------------------------------------------------
   // Visitors of types derived from clang::Decl.
+  bool VisitNamedDecl(clang::NamedDecl* decl) {
+    preprocessor_info()
+        .FileInfoFor(preprocessor_info().main_file())
```

Comment:
This use of `main_file()` was surprising to me. I tried to replace it with `CurrentFileEntry()`, but that caused test failues.

I think I understand now that this is why the namespace info was global before? Because it's statistics over the whole compilation unit, not really keyed by file?

Reply:
To do this the statistics must be recorded per namespace. Under that namespace we subkey by file, allowing us to identify the best file for the namespace.

If we reverse the keys, recording stats on each namespace declared within a file, then we still need something to tell us which files a given namespace is in, or we have to iterate over all files for each namespace we want to find.

So this does look a bit odd, with the `nsinfoPerFile` map is in the `IwyuFileInfo`. If processing multiple translation units in one run, this should make the namespace processing independent for those translation units, which is a good thing.

Reply:
Sorry, I'm not following. You keep saying "this" about something that sounds interesting, but I'm not sure what it is :)

But yeah, the way you put all statistics about namespaces into the main file make me think that moving the state to `IwyuFileInfo` was not the right move.

> If processing multiple translation units in one run, this should make the namespace processing independent
> for those translation units, which is a good thing (not sure if IWYU actually does that).

IWYU only does one translation unit at a time.

But even if it did multiple, I think putting this state into the AST visitor might be a better fit. Let me play around with a refactor for the UsingDecl delayed reporting, I think there may be symmetries/synergies...

Reply:
> Sorry, I'm not following. You keep saying "this" about something that sounds interesting, but I'm not sure what it is :)

My apologies, I've assumed we're on the same page. I guess this PR has been open a while.

> To do this the statistics...

Here I was refering to the PR, 'this' being "finding the best include file for a given namespace".

> So this does look a bit odd...

This instance refers to the current commit. :)

I'll wait for further feedback. I can back up to a version with all the changes except the move into `IwyuFileInfo`. Just let me know your preference.

Reply:
> Let me play around with a refactor for the UsingDecl delayed reporting, I think there may be symmetries/synergies...

Nope, the UsingDecl handling seems slightly different. It seems to keep track of uses within the same file. And it's not remembering uses to replay them later, it has enough information in the UsingDecl to create the uses from scratch. Or at least that's what it does :). Nevermind.

Reply:
FYI, I'm reworking this commit so the `IwyuNamespaceInfo` is stored in `IwyuFileInfo` lists as you might expect. The `nsPerFile` map be static and point to these entries, allowing us to do the lookup during `GetBestNamespaceDecl` as we need to.

Reply:
Sorry for the prolonged silence, I've been alternating between busy with work and vacationing. I'm actually not sure anymore that the namespace statistics should go in `IwyuFileInfo`, given how they're global across files.

I'm a little cautious around static members, because they tend to cause trouble with reentrancy (and threading, which is not an immediate concern here), but let me take a look at your latest patch and see if I can get my head around it.